### PR TITLE
fix concurrently extract file

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
Fix non-atomic concurrent read/write file in #7004 

For example, the size of libcore_worker_library_java.dylib is 13701260. But when return from `org.ray.runtime.util.BinaryFileUtil#getFile`, it may be being written by some other process. Then when load this library, it may be still incomplete., and in this test, it's 13687808, and loadLibrary will fail.
```
./raylet-2020-02-19_16-03-47-06177.out:41:File size 13701120
./raylet-2020-02-19_16-03-47-06177.out:45:File size 13701120
./raylet-2020-02-19_16-03-47-06177.out:47:File size 13687808

(ray) 192:logs chaokunyang$ ls -l ..
total 96072
-rw-r--r--   1 chaokunyang  wheel  13701260 Feb 19 16:03 libcore_worker_library_java.dylib
-rw-r--r--   1 chaokunyang  wheel   2532312 Feb 19 16:03 libray_redis_module.so
-rw-r--r--   1 chaokunyang  wheel  13688256 Feb 19 16:03 libstreaming_java.dylib
```

## Related issue number
#7044

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java [hunk 1]
```diff
@@ -45,4 +47,23 @@ public static File getFile(String destDir, String fileName) {
     }
     return file;
   }
+
+  public static File getFileSafe(String destDir, String fileName) {
```
Comment (jovany-wang) [inline java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java]: `getFile` seems unused now, could you remove it?

**Review by jovany-wang — COMMENTED** [review]
Looks good overall.

File: java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java (hunk 1 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java]: Also comment on this method.

Comment (chaokunyang) [reply java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java]: done

**Review by jovany-wang — APPROVED** [review]
LGTM, thanks for the fix.

File: java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java [hunk 2]
```diff
@@ -22,27 +24,42 @@
...
+  /**
+   * Extract a resource file to <code>destDir</code>.
+   * Note that this a process-safe operation. If multi processes extract the file to same
+   * directory concurrently, this operation will be protected by a file lock.
+   *
+   * @param destDir  a directory to extract resource file to
+   * @param fileName resource file name
+   * @return extracted resource file
+   */
+  public static File getFileSafe(String destDir, String fileName) {
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java]: how about keeping the name `getFile`? I think it's fine as long as "safe" is documented in the comment.

Comment (chaokunyang) [reply java/runtime/src/main/java/org/ray/runtime/util/BinaryFileUtil.java]: fixed

**Review by raulchen — APPROVED** [review]

**Pull request merged**